Changed the default parameter for RegularGridInterpolator

facero opened the pull request:
this avoids the strange effect to have a cross around the spatial
template. With fill_value=None, the default is to extrapolate on the
spatial template pixels values

facero commented:
This found and discussed at HD but apparently no PR were submitted. The cross effect around a spatial template is the following one: 
![download 2](https://user-images.githubusercontent.com/8652898/43087595-82201890-8ea0-11e8-90eb-bf7feea514f0.png)

Changing the default value from `fill_value=None` (default) to `fill_value=0` avoids this issue.

facero commented:
Thanks, 
I looked at the test `test_wcsndmap_interp_by_coord`
and for a reason I'm not sure to understand, it breaks the test if fill_value=0. 
Basically this lines:
```
    m = WcsNDMap(geom)
    coords = m.geom.get_coord(flat=True)
    m.set_by_coord(coords, coords[1])
    assert_allclose(coords[1], m.interp_by_coord(coords, interp='nearest', fill_value=None))
```
 are comparing pixel values filled with interp_by_coord with the coordinates.
with fill_value=0 some values are at 0 while with fill_value=None some there are no 0 as some values are extrapolated.
here is a comparison of the maps `m.set_by_coord`
![m set_by_coord](https://user-images.githubusercontent.com/8652898/43142615-1bb95a76-8f59-11e8-9969-5acb4a25ec2d.png)

and `m.interp_by_coord`
![fill_value0](https://user-images.githubusercontent.com/8652898/43142584-08da02de-8f59-11e8-8700-1a8f5880a1f4.png)

The comparison of the two images explains why the tests crahed ;-)

Should we use fill_value=None (extrapolate) as the default value  and use fill_value=0 when we need it for specific case like in the creation of the spatial template ?

adonath commented:
@facero Thanks for looking into this in detail! Also for backwards compatibility I would suggest to introduce the `fill_value` keyword with `None` as the default value and then set it to different values as needed.

adonath commented on gammapy/maps/tests/test_wcsnd.py:
```diff
@@ -242,9 +242,10 @@ def test_wcsndmap_interp_by_coord(npix, binsz, coordsys, proj, skydir, axes):
+    assert_allclose(coords[1], m.interp_by_coord(coords, interp='nearest', fill_value=None))
+    assert_allclose(coords[1], m.interp_by_coord(coords, interp='linear', fill_value=None))
+    assert_allclose(coords[1], m.interp_by_coord(coords, interp=1, fill_value=None))
+    m.interp_by_coord(coords, interp=1, fill_value=0)
```
Could you use a coordinate that is outside the image and assert that the returned value is zero here?

adonath commented:
I've left again one in-line comment for the test and I think the change in the `SkyDiffuseMap` model is missing, to actually fix the issue you mentioned above. Once these are implemented the PR is ready to merge.

facero commented:
I've added a test to assert zeros with the m.interp_by_coord() from a coord outside the interpolation domain.
I've also added the fill_value in the docstring.
The `SkyDiffuseMap` was modified in a previous commit:
`val = self._map.interp_by_coord(coord, fill_value=0)` in SkyDiffuseMap.
